### ISSUE TITLE
Avoid unnecessary structure updates in renderer

### ISSUE THREAD

**Issue by f1ames** [ckeditor/ckeditor5-engine#1417]
Extracted from #1342:

> Background: #1334.
>
> How does selection conversion work?
>
> 1. MODEL: `<p><$text bold=true>xx</text></p>`  + `selection[bold=true]`
>  2. VIEW:
	1. `<p><b>xx</b></p>`
	2. `<p>[]<b>xx</b></p>` (after position conversion)
	3. `<p><B>[]</B><b>xx</b></p>` (during `writer.wrap( selection )` )
	4. `<p><B>[]xx</B></p>` (after attribute conversion; we're merging to the left, so a new element `<B>` is left)
> 3. RENDERING:
	* actual DOM: `<p><b>xx</b></p>`
	* expected DOM `<p><B>XX</B></p>` (we need to replace the entire `<b>` element; IDEA: don't replace `<b>` with `<B>` because they are similar!)
> 
> BUT... WHAT IF SOMETHING ELSE HAS CHANGED?
> 
> * actual DOM: `<p><b>xx</b></p>`
> * expected DOM `<p><B>XXY</B></p>` (don't replace `<b>` with `<B>` but recursively go deeper to apply other possible changes, see #1125)
> 
> Possible solution: 
> 
> 1. Make `diff( actual, expected, sameNode )` skip elements which have the same names and attributes (see `view.Element#isSimilar()`) - it is described in more details in #1334.
> 2. As a result, we won't replace `<b>` with `<B>`, which means that the view to DOM mapping will be outdated – the new view `<B>` element will be mapped to the new DOM `<B>` element which is not in the tree (because the old DOM `<b>` element was left there). Which means that we should update bindings after skipping such replacements. How to do that? Perhaps we can simply and blindly set new mappings after rendering: purge existing bindings (in a limited scope), `updatedDomElements.forEach( mapToUpdatedViewElement )`.

Also intersting analysis of a specific case is available in https://github.com/ckeditor/ckeditor5-engine/issues/1342#issuecomment-373778804.

**Comment by f1ames** [comment]
To understand what's exactly happening during rendering and how renderer can be improved (so this case is analyzed from [`Renderer`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/renderer.js) point of view), let's take a look how typing inside empty `strong` element is handled.

**Starting with HTML like**:

```
<p>Foo Bar<strong>FILLER{}</strong></p>
```

[`DomConverter`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/domconverter.js) has 3 [mappings](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/domconverter.js#L86) in this state (all elements have existing parent):

1. `RootEditableElement:div` (which points to main editable `div` element)
2. `ContainerElement:p`
3. `AttributeElement:strong`

**Typing `a`**:

1. Mutations are handled via [`MutationObserver`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/observer/mutationobserver.js) and [`input._handleTextNodeInsertion`](https://github.com/ckeditor/ckeditor5-typing/blob/553061fff3c5737c7be3cbca9b61b2f1a698cc2a/src/input.js#L303) method which executes [`input` command](https://github.com/ckeditor/ckeditor5-typing/blob/master/src/inputcommand.js).
1. The `input` command enques changes with [`WeekInsertDelta`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/model/delta/weakinsertdelta.js) (with one [`InsertOperation`](https://github.com/ckeditor/ckeditor5-engine/blob/master/src/model/operation/insertoperation.js) inserting `a` text node).
1. Changes are then processed and converted via [`DowncastDispatcher`](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/conversion/downcastdispatcher.js#L134) which leads to [`writer.wrap`](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/writer.js#L799) handling the character inseriton.
1. The `writer.wrap` internally calls [`writer._wrapRange`](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/writer.js#L1069) as it is handling non-collpased selection.
    * View structure is now `P > text:Foo Bar, text:a, Bold`.
1. Then [`writer._wrapChildren`](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/writer.js#L929) is called which wraps newly inserted character inside new `bold` element
    * View structure is now `P > text:Foo Bar, Bold > text:a, Bold`.
1. On the end of `writer._wrapRange` [attributes are merged](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/writer.js#L1115). Since there were two `bold` elements, the first one is left and the second one is removed.
    * View structure is now `P > text:Foo Bar, Bold > text:a`.
1. `DomConverted` still holds same 3 mappings, however `AttributeElement:strong` has no parent at this point (as it was removed from view structure in the previous step).
1. Rendere has two [elements marked to update](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/renderer.js#L73) now:
    * `AttributeElement:strong` - marked by mutations observer (this is the old element so one without parent).
    * `ContainerElement:p` - marked by view.
1. Since `AttributeElement:strong` (the old one available through `DomConverter` mappings) has no children and real DOM `strong` element has one text child, the text from DOM `strong` element is removed by [`renderer._updateChildren`](https://github.com/ckeditor/ckeditor5-engine/blob/86521df8a00e1ceb90da5852da89eb4f211ca428/src/view/renderer.js#L466).
1. After that, `Renderer` handles `ContainerElement:p` children. Real DOM `strong` element is empty (text was removed in a previous step) and expected one (mapped from view) has text (well it doesn't matter here because diff checks strict element equality and those two are different objects). The new `strong` element is inserted and the old one removed (in such order). Also `DomConverter` mappings are updated here.

Of course there is much more happening during typing, but this steps are important for renderer.

To sum up, transformations performed on the real DOM by the renderer:
1. `<p>Foo Bar<strong>a</strong></p>` (initial DOM state after typing)
1. `<p>Foo Bar<strong></strong></p>` (`a` removed)
1. `<p>Foo Bar<strong>a</strong><strong></strong></p>` (`strong > a` inserted)
1. `<p>Foo Bar<strong>a</strong></p>` (empty `strong` removed, final DOM state)

---

In this particular case `Renderer` basically shouldn't update the DOM as typed character is already inserted by the browser. However, as `Renderer` relies strongly on mappings (`DomConverter`) which are changed by the `Writer` (not directly, just elements are manipulated) and gets outdated due to view structure manipulation it does similar thing as `Writer` inserting new and removing old element. Probably `Renderer` could try match DOM <-> `View` structure smartly in such cases and  introduce only minimal changes (or no changes at all) to the DOM structure and update elements mappings when needed.

**Comment by f1ames** [comment]
I took a look also on case described in https://github.com/ckeditor/ckeditor5-engine/issues/1342#issuecomment-373778804, but now it seems there is only a problem with bold rendering (same as described above) and:

> in `domconverter` we are unable to [map it from an existing view element](https://github.com/ckeditor/ckeditor5-engine/blob/f6de5f5b69947aba95a7ac07731d46772da1cf02/src/view/domconverter.js#L192) so the [new element is created](https://github.com/ckeditor/ckeditor5-engine/blob/f6de5f5b69947aba95a7ac07731d46772da1cf02/src/view/domconverter.js#L216) and then [all children are moved from exisitng DOM element to a newly created one](https://github.com/ckeditor/ckeditor5-engine/blob/f6de5f5b69947aba95a7ac07731d46772da1cf02/src/view/domconverter.js#L229). As the new element is not attached to the DOM all moved children also becomes detached.

no longer occurs. Looks like it was affected by some changes along the way. The `renderer.markedChildren` order is different (now is `strong, a, i, p` and earlier it was `strong, p, a, i`), but I checked with previous order and it also works fine.

But I found similar behaviour when typing on the end of bolded link:
1. `<p>Foo <a><strong>Bar{}</strong></a></p>` (initial DOM state before typing)
1. `<p>Foo <strong><a>Bar</a>a</strong></p>` (initial DOM state after typing - the attributes order swap is native Chrome behaviour)
1. `<p>Foo </p>` (after `renderer` updates `a` children)
1. `<p>Foo <a><strong><a>Bara{}</a></strong></a></p>` (after `renderer` updates `p` children)
1. `<p>Foo <a><strong>Bara{}</strong></a></p>` (after `renderer` updates `strong` children)

This case is different, due to the fact that browser natively changes DOM structure. Here I'm not sure if it would be possible to somehow simplify rendering cycles. The situation here probably could be improved by implementing inline filler improvement mentioned in #1342 (**Fillers** section, point 1).

The fact that whole `strong` is removed in step **3** is caused by the DOM structure changes in step **2**. The DOM `a` element is replaced by the browser in step **2** with the new one so `DomConverter` now holds outdated mappings, still pointing to an old `a` element.
Now the actual DOM children for `a` is empty (as it was modified and detached by the browser) and expected DOM children is the outermost `strong` element.
The `renderer._updateChildren` moves whole `strong` into detached (old) `a` element and so it _disappears_ from DOM.

I also checked Firefox and there it works perfectly. DOM structure is not modified, typed character is inserted on the end of bolded link so `renderer` does not have to perform any changes as expected and actual DOM is exactly the same.

**Comment by f1ames** [comment]
## Bolding selected text

Another interesting case is bolding part of the text. Starting with HTML like:

```
<p>Foo {Bar}</p>
```

and pressing <kbd>ctrl</kbd>/<kbd>cmd</kbd> + <kbd>b</kbd> (or `Bold` button) will result in:

1. `<p>Foo Bar</p>` (initial state)
1. `<p>Foo Foo Bar</p>` (`Foo` text inserted)
1. `<p>Foo <strong>Bar</strong> Foo Bar</p>` (`strong` element inserted)
1. `<p>Foo <strong>Bar</strong</p>` (old `Foo Bar` text removed)

All **2** - **4** steps are happening inside `renderer._updateChildren` (executed for `ContainerElement:p`).

The `renderer` works fine here (and quite optimal) as it needs to insert new bolded element. But now imagine that during collaboration one user is composing something inside `Foo` (e.g. `<p>Fo{}o Bar</p>`) and another one is bolding `Bar` text as described above.

In such case reinserting whole text will break the composition. So the reasonable way would be inserting only `strong` element after existing text and then removing unnecessary part of the text. It is rather an edge case, however main reason this whole issue was brought up was to improve composition handling so I think we should at least think about such cases.

**Comment by f1ames** [comment]
To sum up, as also previously mentioned by @Reinmar (in #1342), it looks like the `renderer` would need a smarter approach when comparing expected and actual DOM so instead of replacing elements which are not identical, smartly reuse them. The part of the solution should be improving [elements diff](https://github.com/ckeditor/ckeditor5-engine/blob/797cd97702c77bc03d39d7de0d16b14adc1a2ed6/src/view/renderer.js#L525):

> 1. Make `diff( actual, expected, sameNode )` skip elements which have the same names and attributes (see `view.Element#isSimilar()`).

which will then require updating mappings:

> 2. As a result, we won't replace `<b>` with `<B>`, which means that the view to DOM mapping will be outdated – the new view `<B>` element will be mapped to the new DOM `<B>` element which is not in the tree (because the old DOM `<b>` element was left there). Which means that we should update bindings after skipping such replacements. How to do that? Perhaps we can simply and blindly set new mappings after rendering: purge existing bindngs (in a limited scope), `updatedDomElements.forEach( mapToUpdatedViewElement )`.

I have some doubts if above will be enough, but it is a good starting point. Such approach should solve most cases similar to the first one (https://github.com/ckeditor/ckeditor5-engine/issues/1417#issuecomment-388809745) described above, when elements are unnecessary replaced even though only its children have changed.

Important thing to keep in mind is that such cases (that new element is already in the DOM) are usually triggered by external actions (like typing). All changes which first appear in `model` and `view` will always need some rerendering, but we should always keep DOM modification minimal. It is especially important during collaboration (and while composing during collaboration). So this solution should also cover these situations.

---

Cases similar to the third one (https://github.com/ckeditor/ckeditor5-engine/issues/1417#issuecomment-389191257) are a little different. Here rendering is needed (as change is triggered internally), but optimizing the number of elements replaced by the `renderer` is important here:

> In such case reinserting whole text will break the composition. So the reasonable way would be inserting only strong element after existing text and then removing unnecessary part of the text. It is rather an edge case, however main reason this whole issue was brought up was to improve composition handling so I think we should at least think about such cases.

At first glance it looks like it may not be solved by the general solution implemented for this issue (mainly because of mixed element / partial text updates) so alternatively we may extract it to a separate issue and handle it after this one.

---

For the cases similar to second one (https://github.com/ckeditor/ckeditor5-engine/issues/1417#issuecomment-389155640 - DOM modified by the browser before rendering), which are more rare and occurs not in all browsers, number of DOM modifications could be optmized. It should be covered (at least partially) by the general approach solution.
If not we may take a closer look on this (**but after implementing first inline filler improvement from #1342**). Still those are rather edge cases, so unless they causes any significant issues I wouldn't treat it as high priority for now.

**Comment by f1ames** [comment]
One more thing which came to my mind is to check if order of rerendering elements makes a difference. As `renderer.markedChildren` are not sorted anywhere the order strictly depends on the particular case. The assumption is that elements could be sorted by the type (`Container`, `Attribute`, etc), so from outermost to most nested (or the opposite).

For the two first cases (3rd one has only one marked element so it is skipped):

**Typing inside empty bold**

1. Default order ( `AttributeElement:strong`, `ContainerElement:p` ), actions:
    * `updateChildren( strong )` - delete
    * `updateChildren( p )` - equal, insert, delete
2. From outermost ( `ContainerElement:p`, `AttributeElement:strong` ):
    * `updateChildren( p )` - equal, insert, delete

`AttributeElement:strong` is not processed in the second case as it was removed in the first step (rerendering `ContainerElement:p`), which means there is basically one DOM operation less.

**Typing on the end of bolded link**

1. Default (`AttributeElement:a`, `ContainerElement:p`, `AttributeElement:strong`):
    * `updateChildren( a )` - insert
    * `updateChildren( p )` - equal, insert
    * `updateChildren( strong )` - delete, delete, insert
2. From innermost (`AttributeElement:a`, `AttributeElement:strong`, `ContainerElement:p`):
    * `updateChildren( a )` - insert
    * `updateChildren( strong )` - delete, delete, insert
    * `updateChildren( p )` - equal, insert
3. From outermost (`ContainerElement:p`, `AttributeElement:a`, `AttributeElement:strong`): 
    * `updateChildren( p )` - equal, insert, delete
    * `updateChildren( a )` - insert
    * `updateChildren( strong )` - equal

Two first orders have exactly the same set of additions/removals, however when outermost element (here `ContainerElement:p`) is processed first (3rd case), the number, of operations decreases.

I think for cases which needs rendering, we should consider sorting elements as it may decrease the number of operations performed on DOM structure.

**Comment by f1ames** [comment]
We have decided to go with `diff` result _post processing_ to try to detect DOM elements which can be reused instead of being replaced. However...

There are some cases in which it is not possible to reuse some existing DOM elements without moving them based on current `diff` results. Consider HTML:

```
<p>1 <b>2</b> <i>3</i> <a>4</a><p>
```

in short the `p` direct children structure can be described as:

```
t1 B t2 I t2 A  // This is actual DOM structure.
```

where `B`, `I`, `T` are attribute elements and `tx` are text nodes (notice that `t2` is used twice as both text nodes have the same content/data so are treated as equal).

Now `<b>2</b>` (with following space) gets removed leaving `<p>1 <i>3</i> <a>4</a><p>` which is

```
t1 I t2 A // This is expected DOM structure.
```

Diffing actual and expected DOM will result in the following set of operation: `equal, delete, insert, equal, delete, delete, delete, insert`. This means:

_operation -> actual DOM structure_
1. `equal:t1` -> `t1 B t2 I t2 A`
1. `delete:B` -> `t1 t2 I t2 A`
1. `insert:I` -> `t1 [I] t2 I t2 A`

Step **3** is the problematic one here. The structure is `t1 t2 I t2 A` and now we need to insert `I` between `t1` and `t2`. It is not possible to reuse actual `I` without moving it (and by moving it we probably lost benefits of not rerendering it - however I haven't yet checked if composition breaks in such cases).
The fact that first `t2` is used (instead of second one which in this case could help) is the result of how `diff` works. But I suppose it doesn't matter that much in general, if `diff` would work in the opposite way (so marking last not first element as equal) we would have exactly same issue in the similar, reversed structure.

For the first solution I suggest skipping such cases (and then see how and if they cause any issues and fix accordingly). This basically means that `diff` result _post processing_ (described in https://github.com/ckeditor/ckeditor5-engine/issues/1417#issuecomment-390615612) will try to match `insert-delete` pairs only between `equal` actions and not in the whole actions set (AFAIR @Reinmar mentioned something similar when we talked F2F).

**Comment by f1ames** [comment]
**TL;DR** The proposed solution refreshes [`DomConverter` bindings](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/domconverter.js#L86) on the beginning of the [`renderer.render` function](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L180) by diffing (using our standard `diff` function) groups of unequal nodes in actual vs expected DOM structure. Based on the refreshed bindings [`renderer._updateChildren`](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L475) will simply reuse DOM elements instead of replacing them with identical ones. Also due to refreshed bindings it is possible to properly [find selection position](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L310) (it was broken earlier due to outdated mappings in some cases).

---

There are a few ideas and reasons behind the way this solution was implemented, most of them already described in this issue, but I wanted to sum it up (will be helpful for PR reviewer at least). As mentioned in a previous comment we decided to try with `diff` result _post processing_ (which finds elements which were replaced in a `View`, but do not have to be replaced in the DOM - just reused). There are three main concepts:

1. `renderer.markedChildren` sorting
1. `domConverter` mappings refreshing
1. `diff` results _post processing_

## `renderer.markedChildren` sorting

The current state is that `renderer` renders elements in an order they were marked to render. As it is based on different events which order is not guaranteed to be always the same (especially in different browsers), the order is quite random. So we do not have control over it. Since with this improvement we would like to reuse the DOM nodes which do not have to be reinserted it makes more sense to start from the outermost elements (also it decreases number of renders in some cases see https://github.com/ckeditor/ckeditor5-engine/issues/1417#issuecomment-389503934).

## `domConverter` mappings refreshing

The first approach was to just refresh the mappings during rendering (so in [`renderer._updateChildren`](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L475)), however as @Reinmar concluded, early mappings refresh [will solve some inline filler issues](https://github.com/ckeditor/ckeditor5-engine/pull/1355#issuecomment-391413097) so we have moved this whole logic on the beginning of the [`render` function](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L180). This allows other mechanisms (like [selection position mapping](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L315)) also take benefit of up-to-date mappings.

## `diff` results _post processing_

This is actually a big part of _`domConverter` mappings refreshing_. The [`diff` function results with the array of actions](https://github.com/ckeditor/ckeditor5-utils/blob/6ef12463c8507c757bd659378ef398a17fccbecc/src/diff.js#L13) (`equal`, `insert`, `delete`) which is then used to transform actual DOM into expected one. There where 2 stages of finding the proper solution:

1. **Adjust [`sameNodes`](https://github.com/ckeditor/ckeditor5-engine/blob/63439481fddb8df054d88c963e8c2ce14283729d/src/view/renderer.js#L525) function so nodes with the same tag name are treated as equal**.

This doesn't work in some cases. Consider the HTML:

```
<p>1</p>
<p>2</p> // this node gets modified to <p>4</p>
<p>3</p>
```

The `diff` function will return `[ equal, insert, delete, equal ]` (so the 2nd `p` is rerendered). The adjusted `diff` will result in `[ equal, equal, insert, delete ]` which means 3rd `p` is reinserted which doesn't make much sense and breaks the rendering.

2. **Manually postprocess `diff` result based on `insert/delete` action pairs**.

This approach assumes that for each group of `insert/delete` actions, similar (same tag name) nodes should be found. For example, `diff` result like `[ insert, delete, equal, insert, insert, delete ]` means there are two `insert/delete` groups which will be processed independently - 1st - `insert, delete` (starting on index 0) and 2nd - `insert, insert, delete` (starting on index 3). This is a valid assumption, however the crucial part is how elements in this groups are compared.

The naive approach is to just much all `insert`s with all `delete`s in an action group and find similar elements. Then replace `insert`/`delete` action with e.g. `replace` (or anything else) which means these element mappings will be rebind.  Works, but not for all cases, consider basic example:

| # | Actual DOM | Expected DOM | `diff` result | postprocessed expected result |
| --- | --- | --- | --- | --- |
| 1 | text Bold | text Bold | insert, **insert**, delete, **delete** | insert, delete, **replace**
| 2 | text Bold | Bold text | **insert**, insert, delete, **delete** | delete, **replace**, insert
| 3 | Bold text | Bold text | **insert**, insert, **delete**, delete | **replace**, insert, delete
| 4 | Bold text | text Bold | insert, **insert**, **delete**, delete | insert, **replace**, delete

Actions in **bold** are the ones which points to similar element (**Bold**) in this case. For Case 1,3,4 it looks like simply replacing found `insert`/`delete` action with `replace` could do the job, for the 2nd case it is not possible (as there is a different order of actions).

The whole postprocessing basically boils down to running `diff` again on `insert`/`delete` groups with assumption that same nodes are the ones with same tag name. And this produces results as expected.

**Comment by scofalik** [comment]
So I read it and tried to understand as much as I could on the first run :P. I think it is promising that you ended up with a double-diffing solution. I had similar kind of a problem in `model.Differ` class, where I also wanted to diff nodes, but in model. The trick there was to differentiate which nodes only changed attributes and which nodes were removed / inserted / did not change. At one point I had a very similar solution - I run two diffs, one of them was getting me insert/delete chains. Then I extracted those insert/delete chains and run the diff second time (with a different diffing function) to see what really changed.

At the end, we have now a different solution in `Differ` (not applicable here) but I am positive about your solution.

**Comment by Reinmar** [comment]
> So the whole postprocessing basically boils down to running diff() again on insert/delete groups with assumption that same nodes are the ones with the same tag name. And this produces expected results.

TBH, I don't understand the last 3 paragraphs of your comment. I guess you're comparing two approaches but you haven't clearly stated that and how does the first one fails exactly. Could you extend that bit because I feel someone may need to get back to this one day.

**Comment by f1ames** [comment]
@Reinmar I modified the description you mentioned, trying to describe more precisely which cases doesn't work, I hope it is easier to understand now.
